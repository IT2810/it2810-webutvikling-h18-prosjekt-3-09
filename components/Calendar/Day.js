import React, { Component } from "react";
import moment from "moment";

import {
  ScrollView,
  View,
  TouchableOpacity,
  Text,
  TextInput
} from "react-native";
import DateTimePicker from "react-native-modal-datetime-picker";
import { Col, Grid } from "react-native-easy-grid";
import { sendNotification } from "../Notification";
import { ID, validateEvent } from "../../utils";
import colors from "../../constants/Colors";
import layout from "../../constants/Layout";
import {
  Button,
  Card,
  List,
  ListItem,
  FormLabel,
  FormInput,
  Header
} from "react-native-elements";

const Day = ({ value, placeholder, onDayChange, today, hasEvent }) => (
  <Button
    disabled={placeholder}
    containerViewStyle={{
      height: layout.window.width / 7,
      width: layout.window.width / 7,
      opacity: placeholder ? 0.2 : 1,
      marginLeft: 0,
      marginRight: 0,
      flex: 1
    }}
    buttonStyle={{
      flex: 1
    }}
    textStyle={{
      color:
        !hasEvent || placeholder
          ? today
            ? colors.light
            : colors.dark
          : colors.light
    }}
    backgroundColor={
      today ? colors.color5 : hasEvent ? colors.color3 : "transparent"
    }
    onPress={() => onDayChange(value)}
    title={moment(value).format("D")}
  />
);

export default Day;

const initialEvent = {
  id: "",
  start: moment().toDate(),
  end: moment()
    .add(1, "hour")
    .toDate(),
  title: ""
};

export class BigDay extends Component {
  state = {
    dateType: "start",
    isDateTimePickerVisible: false,
    ...initialEvent
  };

  showDateTimePicker = dateType =>
    this.setState({ isDateTimePickerVisible: true, dateType });

  hideDateTimePicker = () => this.setState({ isDateTimePickerVisible: false });

  handleDatePicked = date => {
    this.setState({ [this.state.dateType]: date });
    this.hideDateTimePicker();
  };

  handleTitleChange = title => this.setState({ title });

  reset = () => this.setState({ ...initialEvent });

  handleSave = () => {
    const { start, end, title, id } = this.state;

    const isNewEvent = id === "";
    let newEvent = { title, start, end };
    if (isNewEvent) {
      newEvent = {
        ...newEvent,
        id: ID()
      };
    } else {
      newEvent = { ...newEvent, id };
    }

    const invalid = validateEvent(newEvent);
    if (!invalid) {
      if (isNewEvent) {
        this.props.createEvent(newEvent);
      } else {
        this.props.changeEvent(newEvent);
      }
      this.reset();
      sendNotification("success", "Event is created");
    } else {
      sendNotification("error", invalid);
    }
  };

  handleDelete = id => {
    this.props.deleteEvent(id);
    this.reset();
  };

  setEditedEvent = eventId => {
    const { start, end, title } = this.props.events.find(
      ({ id }) => id === eventId
    );
    this.setState({ id: eventId, start, end, title });
  };

  render() {
    const { start, end, dateType, title, id } = this.state;
    const { value, events } = this.props;

    const isNew = id === "";

    return (
      <ScrollView>
        <Header
          backgroundColor={colors.color1}
          centerComponent={{
            text: moment(value).format("MMM DD."),
            style: { color: colors.light }
          }}
          leftComponent={{
            icon: "arrow-back",
            color: "#fff",
            onPress: () => this.props.changeDay(-1)
          }}
          rightComponent={{
            icon: "arrow-forward",
            color: "#fff",
            onPress: () => this.props.changeDay(1)
          }}
        />
        <Grid style={{ marginTop: 16 }}>
          <Col>
            <Button
              backgroundColor={colors.color1}
              onPress={this.handleSave}
              title={isNew ? "Create" : "Save"}
            />
          </Col>
          <Col>
            <Button
              backgroundColor={colors.color4}
              onPress={() =>
                isNew ? this.props.handleClose() : this.handleDelete(id)
              }
              title={isNew ? "Close" : "Delete"}
            />
          </Col>
        </Grid>
        <Card title={`${isNew ? "Create" : "Edit"} event`}>
          <View style={{ flexDirection: "row" }}>
            <FormLabel>Title</FormLabel>
            <FormInput
              placeholder="Enter title"
              onChangeText={this.handleTitleChange}
              value={title}
            />
          </View>
          <Grid>
            <Col>
              <TouchableOpacity
                onPress={() => this.showDateTimePicker("start")}
              >
                <Text>
                  Start date{" "}
                  {start ? moment(start).format("YYYY. MMMM DD") : "not set"}
                </Text>
              </TouchableOpacity>
            </Col>
            <Col>
              <TouchableOpacity onPress={() => this.showDateTimePicker("end")}>
                <Text>
                  End date{" "}
                  {end ? moment(end).format("YYYY. MMMM DD") : "not set"}
                </Text>
              </TouchableOpacity>
            </Col>
          </Grid>
        </Card>
        <Card title="Events on this day:">
          {events.length ? (
            <List>
              {events.map(({ id, title, start, end }) => (
                <ListItem
                  title={title}
                  key={title}
                  subtitle={`${moment(start).format("YYYY-MM-DD")} â€¢ ${moment(
                    end
                  ).format("YYYY-MM-DD")}`}
                  onPress={() => this.setEditedEvent(id)}
                />
              ))}
            </List>
          ) : (
            <Text>No events for today</Text>
          )}
        </Card>
        <DateTimePicker
          date={
            dateType === "start"
              ? moment(start).toDate() || new Date()
              : moment(end).toDate() || new Date()
          }
          isVisible={this.state.isDateTimePickerVisible}
          onConfirm={this.handleDatePicked}
          onCancel={this.hideDateTimePicker}
        />
      </ScrollView>
    );
  }
}
