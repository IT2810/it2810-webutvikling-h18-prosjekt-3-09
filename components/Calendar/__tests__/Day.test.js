import React from "react";
import renderer from "react-test-renderer";
import Day from "../Day";
import moment from "moment";
import colors from "../../../constants/Colors";

describe("Day snapshot", () => {
  let tree;
  let value = new Date();

  test("renders correctly", () => {
    tree = renderer.create(<Day />);
    expect(tree).toMatchSnapshot();
  });

  test("displays day", () => {
    tree = renderer.create(<Day value={value} />);
    expect(tree.toTree().rendered.props.title).toBe(moment(value).format("D"));
  });

  describe("Day type", () => {
    describe("Today", () => {
      describe("day is today", () => {
        beforeAll(() => {
          tree = renderer.create(<Day today={true} />);
        });
        test("today's text color", () => {
          expect(tree.toTree().rendered.props.textStyle.color).toBe(
            colors.light
          );
        });
        test("today's background color", () => {
          expect(tree.toTree().rendered.props.backgroundColor).toBe(
            colors.color5
          );
        });
      });
      describe("day is not today", () => {
        beforeAll(() => {
          tree = renderer.create(<Day />);
        });
        test("not today's text color", () => {
          expect(tree.toTree().rendered.props.textStyle.color).not.toBe(
            colors.light
          );
        });
        test("not today's background color", () => {
          expect(tree.toTree().rendered.props.backgroundColor).not.toBe(
            colors.color5
          );
        });
      });
    });
    describe("Placeholder", () => {
      describe("day is placeholder", () => {
        beforeAll(() => {
          tree = renderer.create(<Day placeholder={true} />);
        });
        test("enables placeholder", () => {
          expect(tree.toTree().rendered.props.disabled).toBe(true);
        });

        test("placeholder's text color", () => {
          expect(tree.toTree().rendered.props.textStyle.color).toBe(
            colors.dark
          );
        });
        test("placeholder's background color", () => {
          expect(tree.toTree().rendered.props.backgroundColor).toBe(
            "transparent"
          );
        });
      });

      describe("day is not placeholder", () => {
        beforeAll(() => {
          tree = renderer.create(<Day placeholder={false} />);
        });
        test("disables placeholder", () => {
          expect(tree.toTree().rendered.props.disabled).toBe(false);
        });
        test("not placeholder's text color", () => {
          expect(tree.toTree().rendered.props.textStyle.color).toBe(
            colors.dark
          );
        });
        test("not placeholder's background color", () => {
          expect(tree.toTree().rendered.props.backgroundColor).toBe(
            "transparent"
          );
        });
      });
    });

    describe("Event", () => {
      describe("day has event", () => {
        beforeAll(() => {
          tree = renderer.create(<Day hasEvent={true} />);
        });
        test("day with event's text color", () => {
          expect(tree.toTree().rendered.props.textStyle.color).toBe(
            colors.light
          );
        });
        test("day with event's background color", () => {
          expect(tree.toTree().rendered.props.backgroundColor).toBe(
            colors.color3
          );
        });
      });

      describe("day has no event", () => {
        beforeAll(() => {
          tree = renderer.create(<Day hasEvent={false} />);
        });
        test("day without event's text color", () => {
          expect(tree.toTree().rendered.props.textStyle.color).not.toBe(
            colors.light
          );
        });
        test("day without event's background color", () => {
          expect(tree.toTree().rendered.props.backgroundColor).not.toBe(
            colors.color3
          );
        });
      });
    });

    // NOTE: Does not actually test anything...?
    describe("Event handling", () => {
      beforeAll(() => {
        tree = renderer.create(<Day onDayChange={jest.fn()} />);
      });

      test("day clicked", () => {
        expect(tree.toTree().rendered.props.onPress()).toBe(undefined);
      });
    });
  });
});
