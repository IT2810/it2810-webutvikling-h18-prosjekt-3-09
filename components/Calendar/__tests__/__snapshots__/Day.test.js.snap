// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day snapshot renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "marginLeft": 15,
        "marginRight": 15,
      },
      undefined,
      Object {
        "flex": 1,
        "height": 107.14285714285714,
        "marginLeft": 0,
        "marginRight": 0,
        "opacity": 1,
        "width": 107.14285714285714,
      },
      undefined,
    ]
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={null}
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      pointerEvents="box-only"
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#9E9E9E",
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 19,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          Object {
            "backgroundColor": "transparent",
          },
          undefined,
          Object {
            "padding": 12,
          },
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          undefined,
          undefined,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {},
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            Array [
              Object {
                "color": "white",
                "fontSize": 20,
              },
              undefined,
              Object {
                "fontSize": 17.5,
              },
              undefined,
              Object {
                "color": "#444",
              },
              undefined,
              undefined,
              undefined,
            ],
          ]
        }
      >
        19
      </Text>
    </View>
  </View>
</View>
`;
